### ISSUE TITLE
Increment NUMBER_BLOCK_NOT_COMPRESSED when !GoodCompressionRatio

### ISSUE THREAD

**Issue by thatsafunnyname** [facebook/rocksdb#4884]
Hello and thanks for RocksDB,

### Expected behavior

NUMBER_BLOCK_NOT_COMPRESSED would be incremented when !GoodCompressionRatio

### Actual behavior

NUMBER_BLOCK_NOT_COMPRESSED is not incremented when !GoodCompressionRatio

### Steps to reproduce the behavior

Use ```ldb load``` to write 10,000 keys where the value is 16,384 bytes, default compression is snappy.  3 different types of values:
  1.) The 16,384 bytes values are all '0's and every data block is compressed.
  2.) The 16,384 bytes values are taken from /dev/urandom and are compression resistant, all fail the GoodCompressionRatio test and none of the data blocks are compressed
  3.) Values alternate between 1.) and 2.) for a half and half mix, half the data blocks are stored compressed and half are not.

### To produce the input for ```ldb load``` and to show the input sizes are the same:

```
1.) All 0's (hex 30)
   
> perl -e '$buf=q{0} x (16*1024);foreach my $k (1..10_000){printf qq{%d ==> 0x%s\n},$k,unpack(q{H*},$buf)}' | wc -c
327798894
> perl -e '$buf=q{0} x (16*1024);foreach my $k (1..10_000){printf qq{%d ==> 0x%s\n},$k,unpack(q{H*},$buf)}' | wc -l
10000

2.) From /dev/urandom

> perl -e 'open($F,q{/dev/urandom});foreach my $k (1..10_000){read($F,$buf,(16*1024));printf qq{%d ==> 0x%s\n},$k,unpack(q{H*},$buf)}' | wc -c
327798894
> perl -e 'open($F,q{/dev/urandom});foreach my $k (1..10_000){read($F,$buf,(16*1024));printf qq{%d ==> 0x%s\n},$k,unpack(q{H*},$buf)}' | wc -l
10000

3.) Half 1.) and half 2.)

> perl -e 'open($F,q{/dev/urandom});foreach my $k (1..10_000){if(0==($k % 2)){read($F,$buf,(16*1024))}else{$buf=q{0} x (16*1024)};printf qq{%d ==> 0x%s\n},$k,unpack(q{H*},$buf)}' | wc -c
327798894
> perl -e 'open($F,q{/dev/urandom});foreach my $k (1..10_000){if(0==($k % 2)){read($F,$buf,(16*1024))}else{$buf=q{0} x (16*1024)};printf qq{%d ==> 0x%s\n},$k,unpack(q{H*},$buf)}' | wc -l
10000
```

An instance of the ```ldb``` tool that has
```
fprintf(stdout,"%s\n",db_->GetOptions().statistics->ToString().c_str());
```
after the compact has run in ```tools/ldb_cmd.cc::DBLoaderCommand::DoCommand()``` and
```
opt.statistics = rocksdb::CreateDBStatistics();
opt.statistics->stats_level_ = StatsLevel::kExceptTimeForMutex;
```
in ```tools/ldb_cmd.cc::DBLoaderCommand::PrepareOptionsForOpenDB()```.

An instance of the ```sst_dump``` tool that has been built with:
```
printf("compression_type=%d\n",compression_type);
```
at around line 226 in ```table/block_fetcher.cc::BlockFetcher::ReadBlockContents```.

### Then the 3 runs:

Note that ```ldb load --compact``` is used so data is written twice.

1.) All 0's (hex 30)
```
> rm -rf /tmp/tmp.rdb ; perl -e '$buf=q{0} x (16*1024);foreach my $k (1..10_000){printf qq{%d ==> 0x%s\n},$k,unpack(q{H*},$buf)}' | ./ldb --create_if_missing --db=/tmp/tmp.rdb --value_hex --compact load |  egrep -i 'rocksdb.number.block.(not_)*compressed'
 
rocksdb.number.block.compressed COUNT : 20004
rocksdb.number.block.not_compressed COUNT : 0
  
> du -shc /tmp/tmp.rdb/*.sst
7.8M    /tmp/tmp.rdb/000013.sst
7.8M    total

> ltrace -e ZSTD_createDCtx-@libzstd.so -e LZ4_createStreamDecode-@liblz4.so* -e _ZN6snappy13RawUncompressEPKcmPc-@libsnappy.co -c ./sst_dump --show_properties  --command=check --file=/tmp/tmp.rdb/000013.sst 2>&1 | egrep -i 'data blocks|entries|compression|ZSTD|calls|LZ4|snappy|average'
  # data blocks: 10000
  # entries: 10000
  raw average key size: 11.889400
  raw average value size: 16384.000000
  SST file compression algo: Snappy
% time     seconds  usecs/call     calls      function
 99.97    0.862004          86     10001 _ZN6snappy13RawUncompressEPKcmPc

> ./sst_dump --file=/tmp/tmp.rdb/000013.sst --command=check | grep compression_type | sort | uniq -c
      4 compression_type=0
  10001 compression_type=1
```

2.) From /dev/urandom

```
> rm -rf /tmp/tmp.rdb ; perl -e 'open($F,q{/dev/urandom});foreach my $k (1..10_000){read($F,$buf,(16*1024));printf qq{%d ==> 0x%s\n},$k,unpack(q{H*},$buf)}' | ./ldb --create_if_missing --db=/tmp/tmp.rdb --value_hex --compact load |  egrep -i 'rocksdb.number.block.(not_)*compressed'

rocksdb.number.block.compressed COUNT : 6
rocksdb.number.block.not_compressed COUNT : 0   (*** NOTE *** )

> du -shc /tmp/tmp.rdb/*.sst
65M     /tmp/tmp.rdb/000013.sst
65M     /tmp/tmp.rdb/000014.sst
29M     /tmp/tmp.rdb/000015.sst
157M    total

> ltrace -e ZSTD_createDCtx-@libzstd.so -e LZ4_createStreamDecode-@liblz4.so* -e _ZN6snappy13RawUncompressEPKcmPc-@libsnappy.co -c ./sst_dump --show_properties  --command=check --file=/tmp/tmp.rdb/000013.sst 2>&1 | egrep -i 'data blocks|entries|compression|ZSTD|calls|LZ4|snappy|average'
  # data blocks: 4090
  # entries: 4090
  raw average key size: 11.889242
  raw average value size: 16384.000000
  SST file compression algo: Snappy
% time     seconds  usecs/call     calls      function
 39.59    0.000192         192         1 _ZN6snappy13RawUncompressEPKcmPc
 
> ./sst_dump --file=/tmp/tmp.rdb/000013.sst --command=check | grep compression_type | sort | uniq -c
   4094 compression_type=0
      1 compression_type=1

> ./sst_dump --file=/tmp/tmp.rdb/000014.sst --command=check | grep compression_type | sort | uniq -c
   4094 compression_type=0
      1 compression_type=1
	  
> ./sst_dump --file=/tmp/tmp.rdb/000015.sst --command=check | grep compression_type | sort | uniq -c
   1824 compression_type=0
      1 compression_type=1
```

3.) Half 1.) and half 2.)

```
> rm -rf /tmp/tmp.rdb ; perl -e 'open($F,q{/dev/urandom});foreach my $k (1..10_000){if(0==($k % 2)){read($F,$buf,(16*1024))}else{$buf=q{0} x (16*1024)};printf qq{%d ==> 0x%s\n},$k,unpack(q{H*},$buf)}' | ./ldb --create_if_missing --db=/tmp/tmp.rdb --value_hex --compact load |  egrep -i 'rocksdb.number.block.(not_)*compressed'

rocksdb.number.block.compressed COUNT : 10005
rocksdb.number.block.not_compressed COUNT : 0   (*** NOTE *** )

> du -shc /tmp/tmp.rdb/*.sst
65M     /tmp/tmp.rdb/000013.sst
19M     /tmp/tmp.rdb/000014.sst
83M     total

> ltrace -e ZSTD_createDCtx-@libzstd.so -e LZ4_createStreamDecode-@liblz4.so* -e _ZN6snappy13RawUncompressEPKcmPc-@libsnappy.co -c ./sst_dump --show_properties  --command=check --file=/tmp/tmp.rdb/000013.sst 2>&1 | egrep -i 'data blocks|entries|compression|ZSTD|calls|LZ4|snappy|average'
  # data blocks: 7796
  # entries: 7796
  raw average key size: 11.888789
  raw average value size: 16384.000000
  SST file compression algo: Snappy
% time     seconds  usecs/call     calls      function
 99.91    0.338492          86      3898 _ZN6snappy13RawUncompressEPKcmPc
 
> ./sst_dump --file=/tmp/tmp.rdb/000013.sst --command=check | grep compression_type | sort | uniq -c
   3903 compression_type=0
   3898 compression_type=1
```

### With https://github.com/facebook/rocksdb/pull/4885

1.) All 0's (hex 30)
```
> rm -rf /tmp/tmp.rdb ; perl -e '$buf=q{0} x (16*1024);foreach my $k (1..10_000){printf qq{%d ==> 0x%s\n},$k,unpack(q{H*},$buf)}' | ./ldb --create_if_missing --db=/tmp/tmp.rdb --value_hex --compact load |  egrep -i 'rocksdb.number.block.(not_)*compressed'

rocksdb.number.block.compressed COUNT : 20004
rocksdb.number.block.not_compressed COUNT : 0  ( No change )
```
2.) From /dev/urandom
```
> rm -rf /tmp/tmp.rdb ; perl -e 'open($F,q{/dev/urandom});foreach my $k (1..10_000){read($F,$buf,(16*1024));printf qq{%d ==> 0x%s\n},$k,unpack(q{H*},$buf)}' | ./ldb --create_if_missing --db=/tmp/tmp.rdb --value_hex --compact load |  egrep -i 'rocksdb.number.block.(not_)*compressed'

rocksdb.number.block.compressed COUNT : 6
rocksdb.number.block.not_compressed COUNT : 20000 
  ( Now records the 10,000 blocks that are not compressed. 
    Once when first written, and then again when compressed)
```
3.) Half 1.) and half 2.)
```
> rm -rf /tmp/tmp.rdb ; perl -e 'open($F,q{/dev/urandom});foreach my $k (1..10_000){if(0==($k % 2)){read($F,$buf,(16*1024))}else{$buf=q{0} x (16*1024)};printf qq{%d ==> 0x%s\n},$k,unpack(q{H*},$buf)}' | ./ldb --create_if_missing --db=/tmp/tmp.rdb --value_hex --compact load |  egrep -i 'rocksdb.number.block.(not_)*compressed'

rocksdb.number.block.compressed COUNT : 10005
rocksdb.number.block.not_compressed COUNT : 10000 
  ( Now records the 5,000 blocks that are not compressed.  
    Once when first written, and then again when compressed)
```

As mentioned in https://github.com/facebook/rocksdb/issues/1474 this is a change in behaviour, but on reflection without this change (or a different change) I could not find a way to expose (to live monitor) the number of blocks that fail the ```GoodCompressionRatio``` check.

Thanks.  PR: ~~https://github.com/facebook/rocksdb/pull/4885~~ https://github.com/facebook/rocksdb/pull/4929

**Comment by anand1976** [comment]
Thanks for reporting this and the PR with the fix

**Comment by thatsafunnyname** [comment]
Merged into master with https://github.com/facebook/rocksdb/commit/79496d71ed9789b33340ba786b83d4a2e8033b00 .
Released with version 6.0.0 - https://github.com/facebook/rocksdb/blob/master/HISTORY.md#600-2192019 
Thanks.
